python/ray/experimental/state/api.py

```diff
@@ -1,195 +1,239 @@
+                see `StateResource` for details.
+            options: List options. See `ListApiOptions` for details.
+            _explain: Print the API information such as API
+                latency or failed query information.
+
+        Returns:
+            A list of queried result from `ListApiResponse`,
+
+        Raises:
+            This doesn't catch any exceptions raised when the underlying request
+            call raises exceptions. For example, it could raise `requests.Timeout`
+            when timeout occurs.
+
+        """
+        # Append the ListApiOptions to the url
+        endpoint = f"/api/v0/{resource.value}?{self._get_query_string(options)}"
+
+        # We don't use `asdict` to avoid deepcopy.
+        # https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict
+        params = {
+            "limit": options.limit,
+            "timeout": options.timeout,
+            "filter_keys": [],
+            "filter_values": [],
+        }
+        for filter in options.filters:
+            filter_k, filter_val = filter
+            params["filter_keys"].append(filter_k)
+            params["filter_values"].append(filter_val)
+
+        # NOTE(rickyyx)
+        # This might raise `requests.Timeout` exceptions and other related requests
+        # exceptions. Users of the functions are expected to handle this.
+        response = self._do_request(
+            "GET",
+            endpoint,
+            timeout=options.timeout,
+            params=params,
+        )
+
+        if response.status_code != 200:
+            self._raise_error(response)
+
+        response = response.json()
+        if response["result"] is False:
+            raise RayStateApiException(
+                "API server internal error. See dashboard.log file for more details. "
+                f"Error: {response['msg']}"
+            )
 
-    if _explain:
         # Print warnings if anything was given.
-        warning_msg = response["data"].get("partial_failure_warning", None)
-        if warning_msg:
-            warnings.warn(warning_msg, RuntimeWarning)
+        warning_msgs = response["data"].get("partial_failure_warning", None)
+        if warning_msgs and _explain:
+            warnings.warn(warning_msgs, RuntimeWarning)
+
+        return response["data"]["result"]
+
+
+"""
+Convenient methods for list_<RESOURCE>
 
-    return r.json()["data"]["result"]
+Supported arguments to the below methods, see `ListApiOptions`:
+    address: The address of the Ray state server. If None, it assumes a running Ray
+        deployment exists and will query the GCS for auto-configuration.
+    timeout: Time for the request.
+    limit: Limit of entries in the result
+"""
 
 
 def list_actors(
-    api_server_url: str = None,
+    address: Optional[str] = None,
     filters: List[Tuple[str, SupportedFilterType]] = None,
     limit: int = DEFAULT_LIMIT,
     timeout: int = DEFAULT_RPC_TIMEOUT,
     _explain: bool = False,
```

Comment:
Wondering if we should factor the warning notification to the user as well. The API could probably return some info w.r.t the warnings (i.e. truncated results) to the user of this function together with the data. 

But didn't do anything in this PR since there is an on-going doc on the API spec. Will follow-up there.

Reply:
Yeah, that was actually one of our original implementations. But I thought it is a bit more natural the return value contains the data directly.

I think this is a pretty important decision to make before we make it alpha (since it will break backward compatibility badly)... Do you mind updating the open question section of this https://docs.google.com/document/d/1eyvSPnYgXBdEXB2-qm_gKDDfO2eEnHe7R1Mu6y-9FWU/edit# to discuss it later?

Reply:
Added: https://docs.google.com/document/d/1eyvSPnYgXBdEXB2-qm_gKDDfO2eEnHe7R1Mu6y-9FWU/edit?disco=AAAAbCnQ1w0